### PR TITLE
Fix Operations Manager + Connection Factory

### PR DESCRIPTION
**Pull request by prolic**
Additional:
- some clean up
- protect against invalid config
- fix demo scripts for clusters

superseeds https://github.com/prooph/event-store-client/pull/35

/cc @pkruithof

### CODE REVIEW

File: src/ConnectionString.php [hunk 1]
```diff
@@ -164,4 +164,14 @@ public static function getUriFromConnectionString(string $connectionString): ?Ur
...
     }
+
+    /**
+     * @param string $connectionString
+     *
+     * @return string[]
+     */
+    private static function getParts(string $connectionString): array
+    {
+        return \explode(';', \str_replace(' ', '', $connectionString));
```
Comment (prolic) [inline src/ConnectionString.php]: This allows 'Connect To' (with spaces)

File: src/EventStoreAsyncConnectionFactory.php [hunk 2]
```diff
@@ -135,6 +135,33 @@ public static function createFromUri(
             );
         }
 
+        if ('' !== $connectionSettings->clusterDns()) {
```
Comment (prolic) [inline src/EventStoreAsyncConnectionFactory.php]: We allow cluster connection via settings only as well.

File: src/Internal/OperationsManager.php [hunk 3]
```diff
@@ -146,6 +146,8 @@ public function checkTimeoutsAndRetry(TcpPackageConnection $connection): void
                 $this->logDebug('retrying, old corrId %s, operation %s', $oldCorrId, $operation);
                 $this->scheduleOperation($operation, $connection);
             }
+
+            $this->retryPendingOperations = [];
```
Comment (prolic) [inline src/Internal/OperationsManager.php]: @pkruithof This is the bad boy for your problem! :)

File: examples/demo-async-cluster.php [hunk 4]
```diff
@@ -17,12 +17,18 @@
...
+ * to the event store in the PHP container, if you run it in Docker.
+ */
+
 Loop::run(function () {
     $builder = new ConnectionSettingsBuilder();
     $builder->setGossipSeedEndPoints([
-        new EndPoint('eventstore1', 2113),
-        new EndPoint('eventstore2', 2123),
-        new EndPoint('eventstore3', 2133),
+        new EndPoint('localhost', 2113),
```
Comment (prolic) [inline examples/demo-async-cluster.php]: @pkruithof I needed to change this in order to make it work. Thoughts?

Comment (pkruithof) [reply examples/demo-async-cluster.php]: It's the way I also was able to connect. It's too bad that you cannot run this from a PHP container yet, but we can work that out later. As long as we're able to test this, I'm fine with it.

File: examples/demo-async-dns-cluster.php [hunk 5]
```diff
@@ -17,13 +17,20 @@
...
+ * This is because the cluster advertises as 127.0.0.1, which does not resolve
+ * to the event store in the PHP container, if you run it in Docker.
+ */
+
 Loop::run(function () {
-    $builder = new ConnectionSettingsBuilder();
-    $builder->setClusterDns('escluster.net');
-    $builder->setClusterGossipPort(2113);
+    $settings = ConnectionSettings::create()
+        ->setClusterDns('127.0.0.1')
```
Comment (pkruithof) [inline examples/demo-async-dns-cluster.php]: I added `escluster.net` to my `hosts` file to test this. It's not a great way and requires an additional manual step, but at least tests the cluster dns stuff.

File: src/Internal/OperationsManager.php (hunk 3 above)
Comment (pkruithof) [reply src/Internal/OperationsManager.php]: I can confirm that my example works with this. I'll test if my real app also works with it (we run it inside a Kubernetes environment with a HA setup)

**Pull request merged**